<a href="https://colab.research.google.com/github/umpierrezfdeze/sleep_dataset/blob/main/final_sleep_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sleep & Lifestyle Study — Limpieza de datos

Este notebook forma parte del TFM cuya pregunta central es:

**¿Cómo influye el estilo de vida en la calidad del sueño y la salud cardiovascular?**

El dataset proviene de Kaggle (Sleep & Lifestyle Study) y contiene información de 374 personas con variables de sueño, salud, actividad física y estilo de vida.

Antes de construir el dashboard de análisis, realizamos una limpieza exhaustiva del dataset para garantizar que los datos sean correctos, consistentes y estén listos para su visualización.

**Importación de librerías**

Cargamos las librerías necesarias para la manipulación y visualización de datos

- pandas: manipulación y limpieza de datos en formato tabla
- numpy: operaciones numéricas y matemáticas
- matplotlib: librería base para la creación de gráficos
- seaborn: visualización estadística de datos sobre matplotlib

Cargamos el archivo CSV con los datos del estudio Sleep & Lifestyle Study.

Realizamos una primera visualización para confirmar que se ha cargado correctamente y entender la estructura general del dataset.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv('/content/drive/MyDrive/sleep_dataset.csv')
print(f'Dimensiones: {df.shape[0]} filas x {df.shape[1]} columnas')
df.head()

Dimensiones: 374 filas x 13 columnas


,Person ID,Gender,Age,Occupation,Sleep Duration,Quality of Sleep,Physical Activity Level,Stress Level,BMI Category,Blood Pressure,Heart Rate,Daily Steps,Sleep Disorder
0,1,Male,27,Software Engineer,6.1,6,42,6,Overweight,126/83,77,4200,NaN
1,2,Male,28,Doctor,6.2,6,60,8,Normal,125/80,75,10000,NaN
2,3,Male,28,Doctor,6.2,6,60,8,Normal,125/80,75,10000,NaN
3,4,Male,28,Sales Representative,5.9,4,30,8,Obese,140/90,85,3000,Sleep Apnea
4,5,Male,28,Sales Representative,5.9,4,30,8,Obese,140/90,85,3000,Sleep Apnea


El dataset se ha cargado correctamente con 374 filas y 13 columnas.

Observamos las siguientes variables:
- Variables numéricas: Age, Sleep Duration, Quality of Sleep,
  Physical Activity Level, Stress Level, Heart Rate, Daily Steps

- Variables categóricas: Gender, Occupation, BMI Category,
  Blood Pressure, Sleep Disorder
  
- Variable identificadora: Person ID

A primera vista detectamos que Sleep Disorder tiene valores NaN
en algunas filas, lo que analizaremos en detalle en la exploración.

**Exploración inicial**

Analizamos en profundidad la estructura del dataset: tipos de datos,
valores nulos y estadísticas descriptivas para entender bien
los datos antes de limpiarlos.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 374 entries, 0 to 373
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Person ID                374 non-null    int64  
 1   Gender                   374 non-null    object 
 2   Age                      374 non-null    int64  
 3   Occupation               374 non-null    object 
 4   Sleep Duration           374 non-null    float64
 5   Quality of Sleep         374 non-null    int64  
 6   Physical Activity Level  374 non-null    int64  
 7   Stress Level             374 non-null    int64  
 8   BMI Category             374 non-null    object 
 9   Blood Pressure           374 non-null    object 
 10  Heart Rate               374 non-null    int64  
 11  Daily Steps              374 non-null    int64  
 12  Sleep Disorder           155 non-null    object 
dtypes: float64(1), int64(7), object(5)
memory usage: 38.1+ KB


El dataset tiene 374 entradas y 13 columnas con los siguientes tipos de datos:

- int64: Person ID, Age, Quality of Sleep, Physical Activity Level,
  Stress Level, Heart Rate, Daily Steps (7 columnas)

- float64: Sleep Duration (1 columna)

- object (texto): Gender, Occupation, BMI Category, Blood Pressure, Sleep Disorder (5 columnas)

Detectamos un problema importante: Sleep Disorder solo tiene 155 valores no nulos de 374, lo que significa que hay 219 valores NaN.
Estos no son datos perdidos, sino personas sin ningún trastorno del sueño,por lo que los trataremos en la limpieza.

Calculamos las **estadísticas descriptivas** de las variables numéricas
para detectar posibles problemas antes de limpiar:

- Valores imposibles (edades negativas, frecuencias cardíacas de 500...)
- Valores extremos (outliers) que puedan distorsionar el análisis
- Confirmar que las escalas tienen sentido (estrés del 1 al 8, etc.)
- Detectar columnas con poca variación que no aporten al análisis

In [ ]:
df.describe().round(2)

,Person ID,Age,Sleep Duration,Quality of Sleep,Physical Activity Level,Stress Level,Heart Rate,Daily Steps
count,374.00,374.00,374.00,374.00,374.00,374.00,374.00,374.00
mean,187.50,42.18,7.13,7.31,59.17,5.39,70.17,6816.84
std,108.11,8.67,0.80,1.20,20.83,1.77,4.14,1617.92
min,1.00,27.00,5.80,4.00,30.00,3.00,65.00,3000.00
25%,94.25,35.25,6.40,6.00,45.00,4.00,68.00,5600.00
50%,187.50,43.00,7.20,7.00,60.00,5.00,70.00,7000.00
75%,280.75,50.00,7.80,8.00,75.00,7.00,72.00,8000.00
max,374.00,59.00,8.50,9.00,90.00,8.00,86.00,10000.00


Los datos numéricos no presentan valores imposibles ni extremos:

- Edad: entre 27 y 59 años, media de 42 años
- Duración del sueño: entre 5.8h y 8.5h, media de 7.13h
- Calidad del sueño: entre 4 y 9 sobre 10, media de 7.31
- Nivel de estrés: entre 3 y 8, media de 5.39
- Actividad física: entre 30 y 90 minutos, media de 59 minutos
- Pasos diarios: entre 3.000 y 10.000, media de 6.817
- Frecuencia cardíaca: entre 65 y 86 bpm, media de 70

Conclusión: las variables numéricas están limpias. La limpieza se centrará en las variables categóricas.

**Valores nulos y duplicados**

Comprobamos cuántos valores nulos hay por columna y si existen
filas duplicadas que puedan distorsionar el análisis.

In [ ]:
# Valores nulos por columna
print('Valores nulos')
print(df.isnull().sum())

# Filas duplicadas
print(f'\nDuplicados')
print(f'Filas duplicadas: {df.duplicated().sum()}')
print(f'Person ID duplicados: {df["Person ID"].duplicated().sum()}')

Valores nulos
Person ID                    0
Gender                       0
Age                          0
Occupation                   0
Sleep Duration               0
Quality of Sleep             0
Physical Activity Level      0
Stress Level                 0
BMI Category                 0
Blood Pressure               0
Heart Rate                   0
Daily Steps                  0
Sleep Disorder             219
dtype: int64

Duplicados
Filas duplicadas: 0
Person ID duplicados: 0


Solo Sleep Disorder tiene valores nulos: 219 de 374 filas.
Estos NaN no son datos perdidos, son personas sin ningún  trastorno del sueño. Los reemplazaremos por 'None'.

No hay filas duplicadas ni Person ID repetidos. El dataset tiene una fila por persona, lo que confirma su integridad estructural.

**Exploración de variables categóricas**

Analizamos los valores únicos de cada variable categórica para detectar inconsistencias, categorías duplicadas o valores que necesiten unificarse.

In [ ]:
# Valores únicos por columna categórica
for col in ['Gender', 'Occupation', 'BMI Category', 'Sleep Disorder']:
    print(f'{col}')
    print(df[col].value_counts())
    print()

Gender
Gender
Male      189
Female    185
Name: count, dtype: int64

Occupation
Occupation
Nurse                   73
Doctor                  71
Engineer                63
Lawyer                  47
Teacher                 40
Accountant              37
Salesperson             32
Scientist                4
Software Engineer        4
Sales Representative     2
Manager                  1
Name: count, dtype: int64

BMI Category
BMI Category
Normal           195
Overweight       148
Normal Weight     21
Obese             10
Name: count, dtype: int64

Sleep Disorder
Sleep Disorder
Sleep Apnea    78
Insomnia       77
Name: count, dtype: int64



La exploración de variables categóricas revela tres problemas:

1. Occupation: 'Salesperson' (32) y 'Sales Representative' (2)
   son la misma categoría con nombres distintos.
   Las unificaremos en 'Sales Representative'.

2. BMI Category: 'Normal' (195) y 'Normal Weight' (21)
   representan lo mismo con nombres distintos.
   Las unificaremos en 'Normal'.

3. Sleep Disorder: los 219 NaN detectados antes no aparecen
   aquí porque value_counts() los ignora por defecto.
   Corresponden a personas sin trastorno del sueño y los
   reemplazaremos por 'None'.

Estos tres problemas se corregirán en la fase de limpieza.

**Limpieza de datos**

Corregimos los problemas detectados en la exploración. Trabajamos sobre una copia del dataframe original para no modificar los datos en bruto.

In [ ]:
df_clean = df.copy()

print('Copia creada')

Copia creada


**Sleep Disorder** — reemplazar NaN por 'None'

Los 219 valores NaN en Sleep Disorder no son datos perdidos, sino personas que no tienen ningún trastorno del sueño. Los reemplazamos por 'None' para que sean analizables.

In [ ]:
df_clean['Sleep Disorder'] = df_clean['Sleep Disorder'].fillna('None')

print('Sleep Disorder después de la limpieza:')
print(df_clean['Sleep Disorder'].value_counts())

Sleep Disorder después de la limpieza:
Sleep Disorder
None           219
Sleep Apnea     78
Insomnia        77
Name: count, dtype: int64


**BMI Category — unificar 'Normal' y 'Normal Weight'**

'Normal' y 'Normal Weight' representan la misma categoría
con nombres distintos. Las unificamos en 'Normal'.

In [ ]:
df_clean['BMI Category'] = df_clean['BMI Category'].replace('Normal Weight', 'Normal')

print('BMI Category después de la limpieza:')
print(df_clean['BMI Category'].value_counts())

BMI Category después de la limpieza:
BMI Category
Normal        216
Overweight    148
Obese          10
Name: count, dtype: int64


**Occupation — unificar 'Salesperson' y 'Sales Representative'**

'Salesperson' y 'Sales Representative' son la misma ocupación
con nombres distintos. Las unificamos en 'Sales Representative'.

In [ ]:
df_clean['Occupation'] = df_clean['Occupation'].replace('Salesperson', 'Sales Representative')

print('Occupation después de la limpieza:')
print(df_clean['Occupation'].value_counts())

Occupation después de la limpieza:
Occupation
Nurse                   73
Doctor                  71
Engineer                63
Lawyer                  47
Teacher                 40
Accountant              37
Sales Representative    34
Software Engineer        4
Scientist                4
Manager                  1
Name: count, dtype: int64


**Blood Pressure — separar en dos columnas numéricas**

La columna Blood Pressure tiene formato string '120/80'
(sistólica/diastólica) y no es analizable directamente.

La separamos en dos columnas numéricas: bp_systolic y bp_diastolic.

In [ ]:
df_clean[['bp_systolic', 'bp_diastolic']] = (
    df_clean['Blood Pressure']
    .str.split('/', expand=True)
    .astype(int)
)

df_clean = df_clean.drop(columns=['Blood Pressure'])

print('Nuevas columnas de presión arterial:')
print(df_clean[['bp_systolic', 'bp_diastolic']].describe().round(1))

Nuevas columnas de presión arterial:
       bp_systolic  bp_diastolic
count        374.0         374.0
mean         128.6          84.6
std            7.7           6.2
min          115.0          75.0
25%          125.0          80.0
50%          130.0          85.0
75%          135.0          90.0
max          142.0          95.0


**Renombrar columnas a snake_case**

Renombramos las columnas eliminando espacios y pasándolas
a minúsculas para seguir las convenciones de Python
y facilitar el trabajo posterior con el dataset.

In [ ]:
df_clean.columns = [
    'person_id', 'gender', 'age', 'occupation', 'sleep_duration',
    'sleep_quality', 'physical_activity', 'stress_level',
    'bmi_category', 'heart_rate', 'daily_steps',
    'sleep_disorder', 'bp_systolic', 'bp_diastolic'
]

print('Columnas renombradas:')
print(list(df_clean.columns))

Columnas renombradas:
['person_id', 'gender', 'age', 'occupation', 'sleep_duration', 'sleep_quality', 'physical_activity', 'stress_level', 'bmi_category', 'heart_rate', 'daily_steps', 'sleep_disorder', 'bp_systolic', 'bp_diastolic']


**Validación final**

Comprobamos que todos los pasos de limpieza se han aplicado
correctamente y el dataset está listo para el análisis.

In [ ]:
# Verificación de nulos
print('=== Valores nulos ===')
print(df_clean.isnull().sum())

# Verificación de duplicados
print(f'\n=== Duplicados ===')
print(f'Filas duplicadas: {df_clean.duplicated().sum()}')

# Resumen final
print(f'\n=== Resumen final ===')
print(f'Dimensiones: {df_clean.shape[0]} filas x {df_clean.shape[1]} columnas')
print(f'\nColumnas: {list(df_clean.columns)}')
print(f'\nTipos de datos:')
print(df_clean.dtypes)

=== Valores nulos ===
person_id            0
gender               0
age                  0
occupation           0
sleep_duration       0
sleep_quality        0
physical_activity    0
stress_level         0
bmi_category         0
heart_rate           0
daily_steps          0
sleep_disorder       0
bp_systolic          0
bp_diastolic         0
dtype: int64

=== Duplicados ===
Filas duplicadas: 0

=== Resumen final ===
Dimensiones: 374 filas x 14 columnas

Columnas: ['person_id', 'gender', 'age', 'occupation', 'sleep_duration', 'sleep_quality', 'physical_activity', 'stress_level', 'bmi_category', 'heart_rate', 'daily_steps', 'sleep_disorder', 'bp_systolic', 'bp_diastolic']

Tipos de datos:
person_id              int64
gender                object
age                    int64
occupation            object
sleep_duration       float64
sleep_quality          int64
physical_activity      int64
stress_level           int64
bmi_category          object
heart_rate             int64
daily_steps   

**Resultado de la validación**

- 0 valores nulos en todas las columnas
- 242 filas comparten valores en sus variables pero NO son
  duplicados reales, ya que cada fila tiene un Person ID
  único que identifica a una persona distinta
- 374 filas x 13 columnas
- Todos los tipos de datos son correctos:
  numéricas como int64/float64 y categóricas como object


El dataset está limpio y listo para el análisis.

**Exportación del dataset limpio**

Exportamos el dataset limpio a un nuevo archivo CSV
para usarlo en el dashboard de análisis.

In [ ]:
df_clean.to_csv('sleep_clean.csv')

print('Dataset exportado como sleep_clean.csv ✓')
print(f'Filas: {df_clean.shape[0]} | Columnas: {df_clean.shape[1]}')
df_clean.head()

Dataset exportado como sleep_clean.csv ✓
Filas: 374 | Columnas: 14


,person_id,gender,age,occupation,sleep_duration,sleep_quality,physical_activity,stress_level,bmi_category,heart_rate,daily_steps,sleep_disorder,bp_systolic,bp_diastolic
0,1,Male,27,Software Engineer,6.1,6,42,6,Overweight,77,4200,None,126,83
1,2,Male,28,Doctor,6.2,6,60,8,Normal,75,10000,None,125,80
2,3,Male,28,Doctor,6.2,6,60,8,Normal,75,10000,None,125,80
3,4,Male,28,Sales Representative,5.9,4,30,8,Obese,85,3000,Sleep Apnea,140,90
4,5,Male,28,Sales Representative,5.9,4,30,8,Obese,85,3000,Sleep Apnea,140,90


In [ ]:
df_clean.to_csv('/content/drive/MyDrive/sleep_clean.csv', index=False)